File: Cloud Storage Samples/Wasabi Browser/cpp/wasabi.cpp
/*
 * Cloud Storage 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#define LINE_LEN 100



class MyWasabi : public Wasabi
{
public:

	virtual int FireObjectList(WasabiObjectListEventParams *e)
	{
		printf("%s\n", e->ObjectName);
		return 0;
	}

	virtual int FireBucketList(WasabiBucketListEventParams *e)
	{
		printf("%s\n", e->BucketName);
		return 0;
	}

	virtual int FireError(WasabiErrorEventParams *e)
	{
		printf("Error Code: %d, Description: %s",e->ErrorCode,e->Description);
		return 0;
	}

};

void printoptions()
{
	printf( "Type ? to select this menu, a number 1-5 to select the following choices:\n"
	        "1.  List Objects\n"
	        "2.  Get Object\n"
	        "3.  Get Object Link\n"
	        "4.  Put Object\n"
	        "5.  Quit\n");
}


int main(int argc, char **argv)
{

	MyWasabi wasabi;
	char buffer[LINE_LEN+1];
	int ret_code = 0;
	int size = 0;

	printf("Wasabi Access Key: ");
	scanf("%80s",buffer);
	wasabi.SetAccessKey(buffer);
	printf("Wasabi Secret Key: ");
	scanf("%80s",buffer);
	wasabi.SetSecretKey(buffer);

	printf("Available buckets:\n");
	if( ret_code = wasabi.ListBuckets() ) goto error;

	printf("\nWasabi Bucket to Examine: ");
	scanf("%80s",buffer);
	wasabi.SetBucket(buffer);

	printoptions();
	while (1)
	{
		printf( "wasabi> " );
		scanf("%80s",buffer);

		if ( ! strcmp(buffer, "?") )
		{
			printoptions();
		}

		else if ( ! strcmp(buffer, "1") )
		{
			if( ret_code = wasabi.ListObjects() ) goto error;
		}

		else if ( ! strcmp(buffer, "2") )
		{
			printf("Which object?: ");
			scanf("%80s",buffer);
			if ( ret_code = wasabi.GetObject(buffer) ) goto error;
			char * data;
			if ( ret_code = wasabi.GetObjectData(data, size) ) goto error;
			printf("Contents of object %s :\n",buffer);
			for(int i=0; i < size; i++)
			{
				printf("%c",data[i]);
			}
			printf("\n");
		}

		else if ( ! strcmp(buffer, "3") )
		{
			printf("Which object?: ");
			scanf("%80s",buffer);
			printf("%s\n", wasabi.GetLink(buffer, 60));
		}

		else if ( ! strcmp(buffer, "4") )
		{
			char objname[81];
			printf("Name of new object: ");
			scanf("%80s",objname);
			printf("Local file to upload: ");
			scanf("%80s",buffer);
			wasabi.SetLocalFile(buffer);
			if( ret_code = wasabi.CreateObject(objname) ) goto error;
			printf("Object %s created.\n",objname);
			wasabi.SetLocalFile("");
		}

		else if ( ! strcmp(buffer, "5") )
		{
			exit(0);
		}

		else if ( ! strcmp(buffer, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Please select a number from [1-5].\n" );
		} // end of command checking
	}  // end of main while loop
	return ret_code;

error:
	printf("\nError: %d", ret_code);
	if (wasabi.GetLastError())
	{
		printf( " \"%s\"\n", wasabi.GetLastError() );
	}
	return ret_code;

}

File: Cloud Storage Samples/Wasabi Browser/java/wasabi.java
/*
 * Cloud Storage 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import cloudstorage.*;

public class wasabi  extends ConsoleDemo {
	
	public wasabi(){
		Wasabi wasabi1 = new Wasabi();
		String buffer= "";
		try{
			wasabi1.addWasabiEventListener(new WasabiEvents());
			System.out.print("Wasabi Access Key: ");
			wasabi1.setAccessKey(input());
			System.out.print("Wasabi Secret Key: ");
			wasabi1.setSecretKey(input());
			System.out.println("\nBucket List:\n");
			wasabi1.listBuckets();
			printMenu();
			while(!buffer.equals("q"))
			{
				System.out.print("Enter command: ");
				buffer = input();
				if(buffer.toLowerCase().equals("cd")){
					System.out.print("Which bucket?: ");
					buffer = input();
					wasabi1.setBucket(buffer);
				}
				else if(buffer.toLowerCase().equals("ls")){
					wasabi1.listObjects();
				}
				else if(buffer.toLowerCase().equals("get")){
					String rem_object;
					System.out.print("Name of object to download: ");
					rem_object = input();
					System.out.print("Directory to download to: ");
					wasabi1.setLocalFile(input() + "\\" + rem_object);
					wasabi1.getObject(rem_object);
				}
				else if(buffer.toLowerCase().equals("put")){
					String rem_object;
					System.out.print("Name of new object: ");
					rem_object = input();
					System.out.print("Local file to upload: ");
					wasabi1.setLocalFile(input());
					wasabi1.createObject(rem_object);
					System.out.println("Object created.");
				}
				else if(buffer.toLowerCase().equals("q")){
					System.exit(0);
				}	
				else if(buffer.toLowerCase().equals("?")){
					printMenu();
				}
			}
		}catch(CloudStorageException e)
        {
            System.out.println(e.getMessage());
            System.exit(e.getCode());
            return;
        }
        catch(Exception e)
        {
            System.out.println(e.getMessage());
        }
		
	}
	
   private void printMenu(){
    	System.out.println("\n?\t-\tHelp\n" +
				"cd\t-\tChange to bucket\n" +
				"ls\t-\tList Objects\n" +
				"get\t-\tGet remoteFile\n" +
				"put\t-\tPut localFile\n" +
				"q\t-\tQuit\n");
    }
    
	public static void main(String[] args) {
		try{
			new wasabi();
		}catch (Exception ex){
			System.out.println("Exception: "+ ex.getMessage());
		}
	}
	
}

class WasabiEvents extends DefaultWasabiEventListener {
	public void bucketList(WasabiBucketListEvent arg) {
		System.out.println(arg.bucketName);
	}
	public void objectList(WasabiObjectListEvent arg) {
		System.out.println(arg.objectName);
	}
	public void SSLServerAuthentication(WasabiSSLServerAuthenticationEvent arg) {
		arg.accept = true;
	}
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}



